src/binder_sim.c

```diff
@@ -0,0 +1,2563 @@
+        ofono_error("Unsupported app type %d", app_type);
+        len = 0;
+    }
+
+    if (len > 0) {
+        hex_path = binder_encode_hex(db_path, len);
+        gbinder_writer_add_cleanup(writer, g_free, hex_path);
+        DBG_(self, "%s", hex_path);
+        return hex_path;
+    } else {
+        /*
+         * Catch-all for EF_ICCID, EF_PL and other files absent
+         * from ef_db table in src/simutil.c, hard-code ROOTMF.
+         */
+        DBG_(self, "%s (default)", ROOTMF);
+        return ROOTMF;
+    }
+}
+
+static
+BinderSimIoResponse*
+binder_sim_io_response_new(
+    const GBinderReader* args)
+{
+    const RadioIccIoResult* result;
+    GBinderReader reader;
+
+    gbinder_reader_copy(&reader, args);
+    result = gbinder_reader_read_hidl_struct(&reader, RadioIccIoResult);
+    if (result) {
+        BinderSimIoResponse* resp = g_slice_new0(BinderSimIoResponse);
+        const char* hex = result->response.data.str;
+
+        DBG("sw1=0x%02X,sw2=0x%02X,%s", result->sw1, result->sw2, hex);
+        resp->sw1 = result->sw1;
+        resp->sw2 = result->sw2;
+        resp->data = binder_decode_hex(hex, -1, &resp->data_len);
+        return resp;
+    }
+    return NULL;
+}
+
+static
+void
+binder_sim_io_response_free(
+    BinderSimIoResponse* res)
+{
+    if (res) {
+        g_free(res->data);
+        g_slice_free(BinderSimIoResponse, res);
+    }
+}
+
+static
+gboolean
+binder_sim_io_response_ok(
+    const BinderSimIoResponse* res)
+{
+    if (res) {
+        static const struct binder_sim_io_error {
+            int sw;
+            const char* msg;
+        } errmsg [] = {
+            /* TS 102.221 */
+            { 0x6a80, "Incorrect parameters in the data field" },
+            { 0x6a81, "Function not supported" },
+            { 0x6a82, "File not found" },
+            { 0x6a83, "Record not found" },
+            { 0x6a84, "Not enough memory space" },
+            { 0x6a86, "Incorrect parameters P1 to P2" },
+            { 0x6a87, "Lc inconsistent with P1 to P2" },
+            { 0x6a88, "Referenced data not found" },
+            /* TS 51.011 */
+            { 0x9240, "Memory problem" },
+            { 0x9400, "No EF selected" },
+            { 0x9402, "Out of range (invalid address)" },
+            { 0x9404, "File id/pattern not found" },
+            { 0x9408, "File is inconsistent with the command" }
+        };
+
+        int low, high, sw;
+
+        switch (res->sw1) {
+        case 0x90:
+            /* '90 00' is the normal completion */
+            if (res->sw2 != 0x00) {
+                break;
+            }
+            /* fall through */
+        case 0x91:
+        case 0x9e:
+        case 0x9f:
+            return TRUE;
+        case 0x92:
+            if (res->sw2 != 0x40) {
+                /* '92 40' is "memory problem" */
+                return TRUE;
+            }
+            break;
+        default:
+            break;
+        }
+
+        /* Find the error message */
+        low = 0;
```

Comment:
This was copied from the old plugin, and I'm quite sure it was done that way because the key is an int and the value is a struct. Yes, one can rely on sw being the first field of the struct and the key always being the first argument to the search function, and that would work, but for whatever reason I feel more comfortable with copy-pasting a binary search template here.